File: scripts/batch_run.pl
#!perl
use strict;
use Getopt::Long;
use MaterialsScript qw(:all);

# List of alloy filenames in your STRUCTURES folder
my @alloys = (
    "Ti15Al1.xsd",
    "Ti14Al2.xsd",
    "Ti12Al4.xsd",
    "Ti8Al8.xsd",
    "Ti4Al12.xsd"
);

print "Starting CASTEP Batch Automation...\n";

foreach my $file (@alloys) {
    # Access the document via its project folder path
    my $docPath = "STRUCTURES/" . $file;
    my $system = $Documents{$docPath};

    if (!defined $system) {
        print "Warning: Could not find $docPath. Skipping...\n";
        next;
    }

    print "Running Geometry Optimization for: $file\n";

    my $results = Modules->CASTEP->GeometryOptimization->Run($system, Settings(
        EnergyCutoffQuality => 'Ultra-fine',
        EnergyCutoff => 420,
        CalculateStress => 'Yes',
        CellOptimization => 'Full',
        CalculateBondOrder => 'Mulliken',
        CalculateCharge => 'Mulliken',
        SpinPolarized => 'No'
    ));

    print "Successfully finished: $file\n";
}

print "All batch calculations completed successfully!\n";

File: scripts/BatchRun.py
import MaterialsScript as MS

project = MS.Project

# 1. Target the folder containing your alloy models
try:
    batch_folder = project.Folders("Structures")
except:
    print("Error: Could not find a folder named 'Structures'.")
    import sys
    sys.exit()

# 2. Configure CASTEP with our verified baseline settings
castep = MS.CASTEP()
castep.Task = "GeometryOptimization"
castep.Functional = "PBE"
castep.CutoffEnergy = 420.0
castep.TreatAsMetal = True
castep.OptimizeCell = True  # Allows the box size to shrink/expand

print("Starting CASTEP Batch Automation...")

# 3. Loop through the folder and run CASTEP sequentially
for item in batch_folder.Items:
    if item.Type == "StructureDocument":
        print("Running Geometry Optimization for: " + item.Name)
        
        # Launch the job using your local CPU
        job = castep.Run(item)
        
        # CRITICAL: Wait for this job to finish before launching the next one
        # This prevents your CPU from crashing by running 10 jobs at once
        job.WaitToCompletion()
        
        print("Successfully finished: " + item.Name)

print("All batch calculations completed successfully!")